File: app/src/main/java/org/ametro/routes/RouteUtils.java
package org.ametro.routes;

import android.support.v4.util.Pair;

import org.ametro.model.entities.MapScheme;
import org.ametro.model.entities.MapSchemeLine;
import org.ametro.model.entities.MapSchemeSegment;
import org.ametro.model.entities.MapSchemeTransfer;
import org.ametro.routes.entities.MapRoute;
import org.ametro.routes.entities.MapRoutePart;

import java.util.HashSet;

public class RouteUtils {

    public static HashSet<Integer> convertRouteToSchemeObjectIds(final MapRoute route, final MapScheme scheme){
        HashSet<Integer> ids = new HashSet<>();

        HashSet<Pair<Integer,Integer>> transfers = new HashSet<>();
        for(MapRoutePart part: route.getParts()){
            ids.add(part.getFrom());
            ids.add(part.getTo());
            transfers.add(new Pair<>(part.getFrom(), part.getTo()));
        }

        for(MapSchemeLine line: scheme.getLines()){
            for(MapSchemeSegment segment: line.getSegments()){
                Pair<Integer,Integer> id = new Pair<>(segment.getFrom(), segment.getTo());
                Pair<Integer,Integer> reverseId = new Pair<>(segment.getTo(), segment.getFrom());
                if(transfers.contains(id) || transfers.contains(reverseId)){
                    ids.add(segment.getUid());
                }
            }
        }

        for(MapSchemeTransfer transfer: scheme.getTransfers()){
            Pair<Integer,Integer> id = new Pair<>(transfer.getFrom(), transfer.getTo());
            Pair<Integer,Integer> reverseId = new Pair<>(transfer.getTo(), transfer.getFrom());
            if(transfers.contains(id) || transfers.contains(reverseId)){
                ids.add(transfer.getUid());
            }
        }

        return ids;
    }

}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/routes/RouteUtils.kt
package io.github.romangolovanov.apps.ametro.routes

import io.github.romangolovanov.apps.ametro.model.entities.MapScheme
import io.github.romangolovanov.apps.ametro.routes.entities.MapRoute

object RouteUtils {

    fun convertRouteToSchemeObjectIds(route: MapRoute, scheme: MapScheme): Set<Int> {
        val ids = mutableSetOf<Int>()
        val transfers = mutableSetOf<Pair<Int, Int>>()

        for (part in route.parts) {
            ids.add(part.from)
            ids.add(part.to)
            transfers.add(part.from to part.to)
        }

        for (line in scheme.lines) {
            for (segment in line.segments) {
                val id = segment.from to segment.to
                val reverseId = segment.to to segment.from
                if (transfers.contains(id) || transfers.contains(reverseId)) {
                    ids.add(segment.uid)
                }
            }
        }

        for (transfer in scheme.transfers) {
            val id = transfer.from to transfer.to
            val reverseId = transfer.to to transfer.from
            if (transfers.contains(id) || transfers.contains(reverseId)) {
                ids.add(transfer.uid)
            }
        }

        return ids
    }
}
